internal/service/networkfirewall/firewall_data_source.go

```diff
@@ -0,0 +1,258 @@
+													Type:     schema.TypeString,
+													Computed: true,
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				AtLeastOneOf: []string{"arn", "name"},
+			},
+			"subnet_change_protection": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"subnet_mapping": {
+				Type:     schema.TypeSet,
+				Computed: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"subnet_id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"tags": tftags.TagsSchema(),
+			"update_token": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"vpc_id": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceFirewallResourceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conn := meta.(*conns.AWSClient).NetworkFirewallConn
+	ignoreTagsConfig := meta.(*conns.AWSClient).IgnoreTagsConfig
+
+	input := &networkfirewall.DescribeFirewallInput{}
+
+	if v, ok := d.GetOk("arn"); ok {
+		input.FirewallArn = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("name"); ok {
+		input.FirewallName = aws.String(v.(string))
+	}
+
+	if input.FirewallArn == nil && input.FirewallName == nil {
+
+		return diag.FromErr(fmt.Errorf("must specify either arn, name, or both"))
+	}
+
+	output, err := conn.DescribeFirewallWithContext(ctx, input)
+	if !d.IsNewResource() && tfawserr.ErrCodeEquals(err, networkfirewall.ErrCodeResourceNotFoundException) {
+		log.Printf("[WARN] NetworkFirewall Firewall (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("error reading NetworkFirewall Firewall (%s): %w", d.Id(), err))
+	}
+
+	if output == nil || output.Firewall == nil {
+		return diag.FromErr(fmt.Errorf("error reading NetworkFirewall Firewall (%s): empty output", d.Id()))
+	}
+
+	firewall := output.Firewall
+
+	d.Set("arn", firewall.FirewallArn)
+	d.Set("delete_protection", firewall.DeleteProtection)
+	d.Set("description", firewall.Description)
+	d.Set("name", firewall.FirewallName)
+	d.Set("encryption_configuration", flattenDataSourceEncryptionConfiguration(firewall.EncryptionConfiguration))
+	d.Set("firewall_policy_arn", firewall.FirewallPolicyArn)
+	d.Set("firewall_policy_change_protection", firewall.FirewallPolicyChangeProtection)
+	d.Set("firewall_status", flattenDataSourceFirewallStatus(output.FirewallStatus))
+	d.Set("subnet_change_protection", firewall.SubnetChangeProtection)
+	d.Set("update_token", output.UpdateToken)
+	d.Set("vpc_id", firewall.VpcId)
+
+	if err := d.Set("subnet_mapping", flattenDataSourceSubnetMappings(firewall.SubnetMappings)); err != nil {
+		return diag.FromErr(fmt.Errorf("error setting subnet_mappings: %w", err))
```

Comment:
You can use `diag.Errorf` instead of `diag.FromErr(fmt.Errorf))`

```suggestion
		return diag.Errorf("error setting subnet_mappings: %s", err)
```